src/server/mod.rs

```diff
@@ -135,6 +136,7 @@ fn find_references<'a>(
 #[derive(Default)]
 struct LspServerState {
     buckets: Vec<String>,
+    composition: Option<RefCell<composition::Composition>>,
```

Comment:
the RefCell is for interior mutation. It should be threadsafe because of the outer mutex lock.